### PR TITLE
Added PostgreNoSQL client to YCSB.

### PR DESCRIPTION
**Pull request by leschekhomann**
Added support for postgre no sql client to YCSB. All commits are squashed.

### CODE REVIEW

**Comment by leschekhomann** [comment]
@busbey: When will I get any feedback regarding my pull request?

**Comment by busbey** [comment]
Hi @leschekhomann! my bandwidth for YCSB stuff is currently tied up in getting a 0.14.0 release out (#1117). If you could help out testing there that would make it easier for me to get time to look at things.

**Comment by leschekhomann** [comment]
Hi @busbey. I was just wondering why I didn‘t get any feedback for such a long time. Now I know the reason. Thanks for the quick reply.

File: postgrenosql/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+**postgrenosql.passwd = postgres**
+
+Defines the password of user the client uses to connect to the Postgres Server.
+
+**postgrenosql.autocommit = true**
+
+Defines whether transactions shoud by applied directly.
+
+
```
Comment (busbey) [inline postgrenosql/README.md]: nit: an example load/run command that uses the properties and/or uses  the example  of a postgrenosql.properties file would help new folks.

File: postgrenosql/src/main/java/com/yahoo/ycsb/db/package-info.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+/**
+ * The YCSB binding for PostgreNoSQL client.
+ */
+package com.yahoo.ycsb.db;
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/db/package-info.java]: please place these classes into a binding-specific package, i.e. `com.yahoo.ycsb.postgresnosql`

File: postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/package-info.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+/**
+ * The YCSB binding for PostgreNoSQL client.
+ */
+package com.yahoo.ycsb.db;
```
Comment (busbey) [inline postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/package-info.java]: wrong package for file location.

File: postgrenosql/pom.xml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+permissions and limitations under the License. See accompanying
+LICENSE file.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>com.yahoo.ycsb</groupId>
+    <artifactId>binding-parent</artifactId>
+    <version>0.14.0-SNAPSHOT</version>
```
Comment (busbey) [inline postgrenosql/pom.xml]: this'll need to be updated now that master is on 0.15.0-SNAPSHOT

File: postgrenosql/src/main/java/com/yahoo/ycsb/db/PostgreNoSQLDBClient.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    String user = props.getProperty(CONNECTION_USER, DEFAULT_PROP);
+    String passwd = props.getProperty(CONNECTION_PASSWD, DEFAULT_PROP);
+    boolean autoCommit = getBoolProperty(props, JDBC_AUTO_COMMIT, true);
+
+    try{
+      Properties tmpProps = new Properties();
+      tmpProps.setProperty("user", user);
+      tmpProps.setProperty("password", passwd);
+
+      postgrenosqlDriver = new Driver();
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/db/PostgreNoSQLDBClient.java]: this method is going to get called on an instance per client thread, so these static variables will need locking to guard them.

A few of the other bindings have examples of doing this, as well as using an AtomicInteger to reference count.

File: postgrenosql/src/main/java/com/yahoo/ycsb/db/PostgreNoSQLDBClient.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+      PreparedStatement readStatement = connection.prepareStatement(createReadStatement(tableName, fields));
+      readStatement.setString(1, key);
+
+      ResultSet resultSet = readStatement.executeQuery();
+      if (!resultSet.next()) {
+        resultSet.close();
+        return  Status.NOT_FOUND;
+      }
+
+      if (result != null && fields != null) {
```
Comment (busbey) [inline postgrenosql/src/main/java/com/yahoo/ycsb/db/PostgreNoSQLDBClient.java]: if `fields` is null, you should include all the record fields in the result.

File: postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClientTest.java [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+  private static Connection postgreSQLConnection = null;
+  private static PostgreNoSQLDBClient postgreNoSQLClient = null;
+
+  @BeforeClass
+  public static void setUp() {
+    // Check whether postgres is available
+    try (Socket socket = new Socket(HOST_NAME, DEFAULT_PORT)){
+      assertThat("Socket is not bound.", socket.getLocalPort(), not(-1));
+    } catch (IOException connectFailed) {
+      assumeNoException("PostgreSQL is not running. Skipping tests.", connectFailed);
```
Comment (busbey) [inline postgrenosql/src/test/java/com/yahoo/ycsb/postgrenosql/PostgreNoSQLDBClientTest.java]: excellent test addition!

Could you update our travis config to add a postgres instance? that'll make sure we get coverage through CI even if individual contributors don't have a local instance running.

https://docs.travis-ci.com/user/database-setup/#PostgreSQL

File: postgrenosql/src/main/java/com/yahoo/ycsb/db/package-info.java (hunk 2 above)
Comment (leschekhomann) [reply postgrenosql/src/main/java/com/yahoo/ycsb/db/package-info.java]: @busbey: It is not really consistent among the different projects. Some add the clients into a new package (Couchbase2) as you suggested, but others (Cassandra, MongoDB) use the com.yahoo.ycsb.db package. In my first implementation I used an own package but changed it due the feedback I received. So what is the right way?

**Pull request closed without merging**